### DISCUSSION TITLE
[Proposal]: `NotNullWhen` on return values

### DISCUSSION THREAD

**Discussion by TobiasKnauss** [dotnet/csharplang#6932] [Language Ideas]
# `NotNullWhen` on return values

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary

The proposed usage of the existing compiler attribute `NotNullWhen` allows declaring a nullable return value as not-null when a boolean parameter has the specified value.

## Motivation
[motivation]: #motivation

In the following example, the returned `XmlNode` may be `null` if the given XPath is not found. If the boolean parameter `i_throw` is `true`, the function will never return `null`, but throw an exception instead. There is no way to tell this to the compiler.

```csharp
public XmlNode? SelectNode (string? i_xPath, bool i_throw = true)
{
  var xmlNode = ...;
  return xmlNode == null && i_throw
           ? throw new Exception ()
           : xmlNode;
}
```

![grafik](https://user-images.githubusercontent.com/18487427/215719476-f4e69822-cbf4-4b99-b777-d02c96f0fd31.png)

With the proposed attribute usage, the compiler would know the behaviour of the function and it would not mark the variable in the screenshot as "may be null".

```csharp
[return: NotNullWhen(nameof(i_throw), true)]
public XmlNode? SelectNode (string? i_xPath, bool i_throw = true)
```

## Detailed design
[design]: #detailed-design

I am not an expert on C# language, so I don't know how the existing implementation must be enhanced or modified, if at all.  
Someone else may 

## Drawbacks
[drawbacks]: #drawbacks

Maybe the benefit is not worth the effort.  
But I don't see any language-related issues.

## Alternatives
[alternatives]: #alternatives

The alternative, which I already use, is creating an overload of the function with return value `XmlNode` instead of `XmlNode?`:

```csharp
public XmlNode SelectNode (string? i_xPath)
{
  var xmlNode = SelectNode (i_xPath, true);
  return xmlNode!;
}

public XmlNode? SelectNode (string? i_xPath, bool i_throw)
{
  ...
}
```

I would like to avoid this extra work.

## Unresolved questions
[unresolved]: #unresolved-questions

**Comment by RikkiGibson** [comment]
Roslyn internally [tends to use](https://github.com/dotnet/roslyn/blob/3a1af4fd52b12c4592878cc6b021bfe0975f554d/src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/DocumentExtensions.cs#L34) a name like `GetRequiredThing` when the throw-on-not-found behavior is desired.

**Comment by jnm2** [comment]
[`Type.GetType`](https://learn.microsoft.com/en-us/dotnet/api/system.type.gettype?view=net-7.0#system-type-gettype(system-string-system-boolean)) and [`Assembly.GetType`](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.assembly.gettype?view=net-7.0#system-reflection-assembly-gettype(system-string-system-boolean)) are more examples I've noticed would play better with NRTs with a facility like this.

**Reply by TobiasKnauss** [reply]
I did not know that .NET FCL methods actuality exist, which use this boolean parameter.

**Comment by olivier-spinelli** [comment]
Many thanks for this proposal. I'm waiting for this too. 

FYI, some people consider that this boolean is a code smell and should be avoided (by using 2 different methods). My humble opinion is that even if it could be done, this inevitably leads to `shouldThrow ? CallThatThrows() : CallThatDoesntThrow()` fragments at places where the behavior HAS to be dynamically chosen (from configuration, from usage context, etc.).

I understand the "code smell" but for low-level API, the boolean helps a lot (at least for me).

**Comment by TobiasKnauss** [comment]
@olivier-spinelli  I am pretty sure that the `throwOnError` parameter is not the only use case for this attribute, it's just the only one that we're discussing so far.

**Reply by olivier-spinelli** [reply]
Totally agree :)

**Comment by dubiousconst282** [comment]
For what is worth, this can be accomplished today with `DoesNotReturnIf`, and I don't think there are edge-cases with it, despite it sounding a bit misleading.

```cs
public Thing? GetThing(string name, [DoesNotReturnIf(true)] bool throwIfNotFound = true) { ... }
```

**Comment by TobiasKnauss** [comment]
@dubiousconst282 I think you are misunderstanding something: The function will of course return if it executed successfully, even if the boolean parameter is true. This parameter only has effect in the case where the function failed to execute: It specifies whether the function will return then or not.
Adding the `DoesNotReturnIf` to the boolean parameter would mean, then the function never returned if this parameter was true, which is wrong.

**Reply by dubiousconst282** [reply]
That's what I meant with it sounding misleading. 

In practice, the `DoesNotReturn` attribute only seems to affect the nullable reachability analysis, so adding it to a function effectively means that it will _never return null_ if the argument has the specified value. This behavior is unlikely to change based on https://github.com/dotnet/csharplang/issues/2555#issuecomment-495694037 and https://github.com/dotnet/csharplang/issues/3826#issuecomment-681191954.

**Reply by TobiasKnauss** [reply]
Yes, it is true that it will _never return null_ if the parameter is true and the attribute exists. This would indeed meet the target.
But it also means that it _never returns at all_, which might impact the nullability analysis and other analysis inside the calling function in a negative way...? Therefore I am not sure whether it meets the target completely or only partially and with negative side effects.

**Reply by RikkiGibson** [reply]
The main concern I'd have is, it might make the compiler assume that expressions after the call are in an *unreachable* state where everything is assumed to be non-null.

**Reply by dubiousconst282** [reply]
Did a quick test, and indeed using `DoesNotReturnIf` causes the issues mentioned. The analysis will stop at the call and every other value will be considered to be non-null:

```cs
void M() {
    string? a = null;
    string? b = B(true);

    a.Contains('1'); //no warning for `a`
}
string? B([DoesNotReturnIf(true)] bool r) => null;
```

**Reply by TobiasKnauss** [reply]
@dubiousconst282
> The analysis will stop at the call and every other value will be considered to be non-null

I rather think that the analysis will stop at the call and nothing else will be considered at all. There simply will be no further checks inside this method, hence no warnings, because `DoesNotReturnIf` tells the analysis that the rest of the method is unreachable.

I appreciate your idea, but as we found out, it looks like it does not produce the desired result. I think we still need this proposal.

**Comment by hughesjs** [comment]
Even just simple:
```
[return: NotNullIfTrue(paramName)]
[return: NotNullIfFalse(paramName)] 
```
Would be a huge improvement and probably easier to implement than the more general `[NotNullIfTrue(expr)]`

**Comment by Thomas-Shephard** [comment]
This would be amazingly helpful as it would allow better support for methods where you either throw an exception if the item is not found or return null based on a parameter.

I think it would make sense to me for the attribute to be used like the DoesNotReturnIfAttribute i.e. NotNullIf on a parameter so that could be used like this.

```csharp

internal static string? GetAttribute(this XmlReader reader, string attributeName, [NotNullIf(true)] bool throwIfNotFound = true) {
    string? attributeValue = reader.GetAttribute(attributeName);

    if (throwIfNotFound && attributeValue is null)
        throw new Exception($"Attribute '{attributeName}' not found");

    return attributeValue;
}
```

**Comment by CaptaiNiveau** [comment]
I've often thought that something like this exists, went to documentation only to remember that it doesn't.
Would love to see this added :)

**Comment by logiclrd** [comment]
Yes, please. :-)

**Comment by TonyValenti** [comment]
There is a simple solution to this:
Create two methods:
```
T? TryGetValue(...) //Does not throw if not found
T GetValue(...) //Throws if not found
```

**Reply by olivier-spinelli** [reply]
The point IS to avoid the two methods: see https://github.com/dotnet/csharplang/discussions/6932#discussioncomment-4851973

**Reply by logiclrd** [reply]
Also, the condition isn't "is it found or not?". The condition is, "does this parameter have this particular value or not?". The two methods apply only in one very specific pattern.

**Comment by wiz0u** [comment]
Please add
```csharp
[return: NotNullIfTrue(paramName)]
[return: NotNullIfFalse(paramName)] 
```
as suggested by @hughesjs 

It's simple, clear, and solves many scenario already

**Comment by bkqc** [comment]
I have a somewhat different use case.
I have an helper method that will format a string based on some criteria and that could return either null or string.empty if criteria are not met. I agree I could add an overload with OrNull but it already exists in the API and being able to annotate it correctly would help greatly.
